### PR TITLE
Performance best practice update with davidfowl samples

### LINKED ISSUE #14605: Migrate AspNetCoreDiagnosticScenarios/AspNetCoreGuidance.md
Move 
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AspNetCoreGuidance.md under **Test, debug, and troubleshoot** 
image

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #14605
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#avoid-using-synchronous-readwrite-overloads-on-httprequestbody-and-httpresponsebody)

### CODE REVIEW

**Comment by davidfowl** [comment]
@Rick-Anderson where's the review url?

**Comment by Rick-Anderson** [comment]
@davidfowl  This is WIP. I forgot to do the draft when I created the PR. I'm just getting started, not ready for review.

[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734)

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs [hunk 1]
```diff
@@ -15,8 +15,6 @@ public async void Get()
         {
             await Task.Delay(1000);
 
-            // THIS will crash the process because it's writing after the response has
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs]: Why did you remove this comment?

**Comment by Rick-Anderson** [comment]
@davidfowl  ready for review at [Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#avoid-using-synchronous-readwrite-overloads-on-httprequestbody-and-httpresponsebody)

Please search for `<!-- TODO Review`  I have questions on those two sections.

* Moved all code comments out of code and into the document. Code comments are not MT (Machine Translated).
* Changed code to get it to compile. For example you use `app.Use(async (next, context) =>` and I changed that to `app.Use(async (context, next) =>`
* Shortened sentences so they would MT. Even native readers can comprehend more with shorter sentences.

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs (hunk 1 above)
Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs]: > Why did you remove this comment?


See my  notes below. We generally move all comments to the doc so they can be MT.

Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs]: Let me know which ones you want back and I can bend the rules and add a few.

Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncBadVoidController.cs]: Edited and added back :)

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            return SearchResults.Combine(results1, results2, results3);
+        }       
+
+        private async Task<SearchResults> SearchAsync(SearchEngine engine, string query)
+        {
+            var searchResults = _searchService.Empty();
+            try
+            {
+                _logger.LogInformation("Starting search query from {path}.", 
+                                                             HttpContext.Request.Path);
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: The formatting here is weird. Can we fix it?

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+            return await searchResults;
+        }
+        #endregion
+
+        private readonly ILogger<AsynchBadSearchController> _logger;
+        public readonly SearchService _searchService;
+
+        public AsynchBadSearchController(ILogger<AsynchBadSearchController> logger,
+                                                        SearchService searchService)
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: The formatting here is weird too. Can we left align the parameters?

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    #region snippet2
+    public class AsyncGoodSearchController : Controller
+    {       
+        [HttpGet("/search")]
+        public async Task<SearchResults> Get(string query)
+        {
+            string path = HttpContext.Request.Path;
+            var query1 = SearchAsync(SearchEngine.Google, query,
+                path);
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: Same here?

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        }
+
+    }
+
+    public class SearchEngine
+    {
+        public static SearchEngine Bing { get; internal set; }
+        public static SearchEngine Google { get; internal set; }
+        public static SearchEngine DuckDuckGo { get; internal set; }
+
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: nit: Remove extra space

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        private readonly ILogger<AsyncGoodSearchController> _logger;
+        public readonly SearchService _searchService;
+
+        public AsyncGoodSearchController(ILogger<AsyncGoodSearchController> logger,
+                                                         SearchService searchService)
+        {
+            _logger = logger;
+            _searchService = searchService;
+        }
+
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: nit: Remove extra space

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            throw new NotImplementedException();
+        }
+
+        public static implicit operator Task<object>(SearchResults v)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: nit: Remove extra space

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        public void Add (Pokemon p) { }
+    }
+
+    public class PokemonDbContext
+    {
+        public  Pokemon  Pokemon;
+
+        internal Task SaveChangesAsync()
+        {
+            throw new NotImplementedException();
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs]: Make this return Task.CompletedTask

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MyFirstController.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    }
+    #endregion
+
+    #region snippet2
+    public class GoodStreamReaderController : Controller
+    {
+        [HttpGet("/pokemon")]
+        public async Task<ActionResult<PokemonData>> Get()
+        {
+            var json = await new StreamReader(Request.Body).ReadToEndAsync();
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MyFirstController.cs]: Since we have an async JSON serializer, lets change this to:

```C#
return await JsonSerializer.DeserializeAsync<PokemonData>(Request.Body);
```

File: aspnetcore/performance/performance-best-practices/samples/3.0/Program.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        public static IHostBuilder CreateHostBuilder(string[] args) =>
+            Host.CreateDefaultBuilder(args)
+             .ConfigureLogging(logging =>
+             {
+                 logging.ClearProviders();
+                 logging.AddConsole();
+             })
+                .ConfigureWebHostDefaults(webBuilder =>
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Program.cs]: nit: Fix formatting here to align the code properly

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+            }
+
+            return await searchResults;
+        }
+        #endregion
+
+        private readonly ILogger<AsynchBadSearchController> _logger;
+        public readonly SearchService _searchService;
+
+        public AsynchBadSearchController(ILogger<AsynchBadSearchController> logger,
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/AsyncFirstController.cs]: Remove the h from `Asynch`

File: aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+            app.UseAuthorization();
+
+            app.UseEndpoints(endpoints =>
+            {
+                endpoints.MapControllers();
+            });
+        }
+    }
+
+    public class SearchService : IDisposable {
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs]: What's this for?

Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs]: > What's this for?
Auto code generation by VS to get the code to compile. It offered a fix and I clicked "OK". Please suggest better code.

Comment (davidfowl) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs]: Remove it all and just implement the interface 😄 .

Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Startup.cs]: `public class SearchService : IDisposable {`

> What's this for?

VS generated code so your snippet would compile.

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MyFirstController.cs (hunk 9 above)
Comment (davidfowl) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MyFirstController.cs]: Remove the StreamReader logic.

File: aspnetcore/performance/performance-best-practices.md [hunk 13]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+The preceding code frequently captures a null or incorrect `HttpContext` in the constructor.
+
+**Do this:** The following example:
+
+* Stores the <xref:Microsoft.AspNetCore.Http.IHttpContextAccessor> in a field.
+* Uses the `HttpContext` field at the correct time and checks for `null`.
+
+[!code-csharp[](performance-best-practices/samples/3.0/MyType.cs?name=snippet2)]
+
+## Do not access HttpContext from multiple threads
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices.md]: The highlights are messed up in this section.

image

File: aspnetcore/performance/performance-best-practices.md [hunk 14]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/MyFirstController.cs?name=snippet1)]
+
+In the preceding code, `Get` synchronously reads the entire HTTP request body into memory. If the client is slowly uploading, the app is doing sync over async. The app does sync over async because Kestrel does **NOT** support synchronous reads.
+
+**Do this:** The following example uses <xref:System.IO.StreamReader.ReadToEndAsync*> and does not block the thread while reading.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/MyFirstController.cs?name=snippet2)]
+
+The preceding code asynchronously reads the entire HTTP request body into memory.
```
Comment (davidfowl) [inline aspnetcore/performance/performance-best-practices.md]: If we remove the ReadToEndAsync and show the async JSON reading directly then it's fine to remove this

File: aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+                await next();
+
+                context.Response.Headers["test"] = "test value";
+            });
+            #endregion
+
+            #region snippet2
+            app.Use(async (context, next) =>
+            {
+                await context.Response.WriteAsync("Hello ");
```
Comment (halter73) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs]: Remove this line. With this here, it makes no sense why we're checking `context.Response.HasStarted` since it would almost\* always be true. If only `next()` is called, it remains ambiguous whether or not the response started.

\* `context.Response.HasStarted` could still be false if the response body Stream was replaced before attempting to write to it, but since it's not mentioned anywhere that may have happened, I'd just remove the attempted write.

File: aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+        public void ConfigureServices(IServiceCollection services)
+        {
+        }
+
+        public void Configure(IApplicationBuilder app)
+        {
+            #region snippet1
+            app.Use(async (context, next) =>
+            {
+                await context.Response.WriteAsync("Hello ");
```
Comment (halter73) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Startup22.cs]: I'd remove this line too. Even though this is the "bad" example, it's supposed to be bad because we're not checking `context.Response.HasStarted`, not just because we wrote to the response body and then tried changing the response header (which is also bad ofc).

File: aspnetcore/performance/performance-best-practices.md (hunk 13 above)
Comment (davidfowl) [reply aspnetcore/performance/performance-best-practices.md]: it's still busted https://review.docs.microsoft.com/en-us/aspnet/core/performance/performance-best-practices?view=aspnetcore-3.1&branch=pr-en-us-14734#do-not-access-httpcontext-from-multiple-threads

**Comment by davidfowl** [comment]
@Rick-Anderson can we rename the section to:

Performance and Reliability?

File: aspnetcore/performance/performance-best-practices.md [hunk 17]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+Naively storing a large request or response body into a single `byte[]` or `string`:
+
+* May result in quickly running out of space in the LOH.
+* May cause performance issues for the app because of full GCs running.
+
+## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing
+
+When using a serializer/de-serializer that only supports synchronous reads and writes:
+
+* Choose to buffer the data into memory before passing data into the serializer/de-serializer.
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]:
```suggestion
* Buffer the data into memory asynchronously before passing it into the serializer/de-serializer.
```

File: aspnetcore/performance/performance-best-practices.md [hunk 18]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+* [JSON.NET](https://www.newtonsoft.com/json/help/html/Introduction.htm) is only synchronous.
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]: This bullet point is in the wrong place. It's an example of a sync serializer, not an action to take when using one.

File: aspnetcore/performance/performance-best-practices.md (hunk 13 above)
Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices.md]: It's correct on edge/FF and Chrome at 80%. The build team (who writes the HTML generator for all of doc.ms.com) broke it 6 months ago and have only managed to mostly fix it.

Edge:
image

FF:
image

Chrome at 80%:

image

File: aspnetcore/performance/performance-best-practices.md [hunk 19]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+This [blog post](https://adamsitnik.com/Array-Pool/#the-problem) describes the problem succinctly:
+
+> When a large object is allocated, it’s marked as Gen 2 object. Not Gen 0 as for small objects. The consequences are that if you run out of memory in LOH, GC cleans up the whole managed heap, not only LOH. So it cleans up Gen 0, Gen 1 and Gen 2 including LOH. This is called full garbage collection and is the most time-consuming garbage collection. For many applications, it can be acceptable. But definitely not for high-performance web servers, where few big memory buffers are needed to handle an average web request (read from a socket, decompress, decode JSON & more).
+
+Naively storing a large request or response body into a single `byte[]` or `string`:
+
+* May result in quickly running out of space in the LOH.
+* May cause performance issues for the app because of full GCs running.
+
+## Use buffered and synchronous reads and writes as an alternative to asynchronous reading and writing
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]: Working with a synchronous data processing API

File: aspnetcore/performance/performance-best-practices.md [hunk 20]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+
+When using a serializer/de-serializer that only supports synchronous reads and writes:
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]: like JSON.NET...

File: aspnetcore/performance/performance-best-practices.md [hunk 21]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2)]
+
+The following highlighted code:
+
+* Creates a scope for the lifetime of the background operation and resolves services from it.
+* Uses `PokemonDbContext` from the correct scope.
+
+[!code-csharp[](performance-best-practices/samples/3.0/Controllers/FireAndForgetSecondController.cs?name=snippet2&highlight=9-16)]
+
+## Avoid adding headers after the HttpResponse has started
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]:
```suggestion
## Do not modify the status code or headers after the response body has started
```

File: aspnetcore/performance/performance-best-practices.md [hunk 22]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+ASP.NET Core does not buffer the HTTP response body. The first time the response is written:
+
+* The headers are sent along with that chunk of the body to the client.
+* It's no longer possible to change response headers.
+
+**Do not do this:** The following code tries to add response headers after the response has already started:
+
+[!code-csharp[](performance-best-practices/samples/3.0/Startup22.cs?name=snippet1)]
+
+In the preceding code, `context.Response.Headers["test"] = "test value";` if `next()` has written to the response.
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]:
```suggestion
In the preceding code, `context.Response.Headers["test"] = "test value";` will throw an exception if `next()` has written to the response.
```

File: aspnetcore/performance/performance-best-practices.md (hunk 21 above)
Comment (Tratcher) [reply aspnetcore/performance/performance-best-practices.md]: Related: Do not call `next()` if you have already started writing to the response body. Components only expect to be called if it's possible for them to handle/manipulate the response.

File: aspnetcore/performance/performance-best-practices.md [hunk 23]
```diff
@@ -132,19 +131,200 @@ Recommendations:
...
+
+**Do this:** The following example checks if the HTTP response has started before writing to the body.
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices.md]:
```suggestion
**Do this:** The following example checks if the HTTP response has started before modifying the headers.
```

File: aspnetcore/performance/performance-best-practices.md (hunk 13 above)
Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices.md]: @davidfowl  see #14936
I asked for the problem to be  prioritized on the internal bug.

File: aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MySecondController.cs [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+    [Route("api/[controller]")]
+    [ApiController]
+    #region snippet1
+
+    public class BadReadController : Controller
+    {
+        [HttpPost("/form-body")]
+        public IActionResult Post()
+        {
+            var form =  HttpContext.Request.Form;
```
Comment (Tratcher) [inline aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MySecondController.cs]: This is misleading to show in a controller, MVC usually reads the form for you async during model binding.

Comment (Rick-Anderson) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MySecondController.cs]: @Tratcher  this comes from https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AspNetCoreGuidance.md#prefer-using-httprequestreadasformasync-over-httprequestform
I'm not allowed to change code so you'll have to provide new code.

Comment (davidfowl) [reply aspnetcore/performance/performance-best-practices/samples/3.0/Controllers/MySecondController.cs]: We can change it to middleware but the point is to show both

**Comment by davidfowl** [comment]
@Rick-Anderson the highlighting in the "Do not access HttpContext from multiple threads" is still busted but I think after it's fixed, we should merge this.
